language/move-bytecode-verifier/src/loop_summary.rs

```diff
@@ -0,0 +1,276 @@
+        exploration.insert(root_block, InProgress(root_node));
+
+        let mut stack: Vec<Frontier> = cfg
+            .successors(root_block)
+            .iter()
+            .map(|succ| Visit {
+                from_node: root_node,
+                to_block: *succ,
+            })
+            .collect();
+
+        while let Some(action) = stack.pop() {
+            match action {
+                Finish {
+                    block,
+                    node_id,
+                    parent,
+                } => {
+                    descs[usize::from(parent)] += 1 + descs[usize::from(node_id)];
+                    *exploration.get_mut(&block).unwrap() = Done(node_id);
+                }
+
+                Visit {
+                    from_node,
+                    to_block,
+                } => match exploration.entry(to_block) {
+                    Entry::Occupied(entry) => match entry.get() {
+                        // Cyclic back edge detected by re-visiting `to` while still processing its
+                        // children.
+                        InProgress(to_node) => backs[usize::from(*to_node)].push(from_node),
+
+                        // Cross edge detected by re-visiting `to` after it and its children have
+                        // been processed.
+                        Done(to_node) => preds[usize::from(*to_node)].push(from_node),
+                    },
+
+                    // Visiting `to` for the first time: `from` must be its parent in the depth-
+                    // -first spanning tree, and we should continue exploring its successors.
+                    Entry::Vacant(entry) => {
+                        let to_node = next_node.bump();
+                        entry.insert(InProgress(to_node));
+                        blocks[usize::from(to_node)] = to_block;
+                        preds[usize::from(to_node)].push(from_node);
+
+                        stack.push(Finish {
+                            block: to_block,
+                            node_id: to_node,
+                            parent: from_node,
+                        });
+
+                        stack.extend(cfg.successors(to_block).iter().map(|succ| Visit {
+                            from_node: to_node,
+                            to_block: *succ,
+                        }));
+                    }
+                },
+            }
+        }
+
+        LoopSummary {
+            blocks,
+            descs,
+            backs,
+            preds,
+        }
+    }
+
+    /// Decides whether `descendant` is a descendant of `ancestor` in the depth-first spanning
+    /// tree.
+    pub fn is_descendant(&self, NodeId(ancestor): NodeId, NodeId(descendant): NodeId) -> bool {
+        // All the descendants of `ancestor` in the DFST will have the IDs immediately following
+        // it.
+        ancestor <= descendant && descendant <= ancestor + self.descs[ancestor as usize]
```

Comment:
> If you had something like `descendant <= max_descendent(ancestor)` that would make sense to me, but it doesn't feel like that's what this check is

That is what this check is, i.e. `max_descendant(ancestor) = ancestor + self.descs[ancestor as usize]`.  We get this nice property because of what the comment above this line mentions:

```
// All the descendants of `ancestor` in the DFST will have the IDs immediately following
// it.
```

This, combined with the the fact that `self.descs[n]` is the number of transitive descendants of `n` gives us everything we need.